language/move-compiler/src/naming/ast.rs

```diff
@@ -361,15 +374,35 @@ impl BuiltinTypeName_ {
             B::Address | B::U8 | B::U64 | B::U128 | B::Bool => AbilitySet::primitives(loc),
             B::Signer => AbilitySet::signer(loc),
             B::Vector => AbilitySet::collection(loc),
+            B::Fun => AbilitySet::functions(),
         }
     }
 
-    pub fn tparam_constraints(&self, _loc: Loc) -> Vec<AbilitySet> {
+    pub fn tparam_constraints(&self, _loc: Loc, arity: usize) -> Vec<AbilitySet> {
         use BuiltinTypeName_ as B;
         // Match here to make sure this function is fixed when collections are added
         match self {
             B::Address | B::Signer | B::U8 | B::U64 | B::U128 | B::Bool => vec![],
             B::Vector => vec![AbilitySet::empty()],
+            B::Fun => (0..arity).map(|_| AbilitySet::empty()).collect(),
+        }
+    }
+
+    pub fn variance(&self, pos: usize, arity: usize) -> Variance {
+        match self {
+            // Function variance: given g: T1 -> R1 and f: T2 -> R2, then
+            // f can substitute g if T2 >= T1 && R1 <= R2
+            BuiltinTypeName_::Fun if pos < arity - 1 => Variance::ContraVariant,
```

Comment:
```suggestion
            BuiltinTypeName_::Fun if pos + 1< arity => Variance::ContraVariant,
```

to avoid underflow if `arity` is 0?

Reply:
That is the standard rule for variance of function types. Input arguments are contravariant, output arguments covariant. Variance is about substitutability/assignability: you can replace with a subtype for covariance, and with a supertype for contra-variance.  It is needed for the `join` function.

If you want to treat function types just as regular type constructors, what I'm doing here (and is common in multiple languages e.g. C# and I bet also internally in Rust), you need to introduce the concept of variance to define the subtype relation.